runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/SplittableParDoExpander.java

```diff
@@ -214,39 +222,124 @@ public MessageWithComponents getReplacement(
             generateUniqueId(
                 transformId + "/ProcessSizedElementsAndRestrictions",
                 existingComponents::containsTransforms);
-        {
-          PTransform.Builder processSizedElementsAndRestrictions = PTransform.newBuilder();
-          processSizedElementsAndRestrictions.putInputs(mainInputName, splitAndSizeOutId);
-          processSizedElementsAndRestrictions.putAllInputs(sideInputs);
-          processSizedElementsAndRestrictions.putAllOutputs(splittableParDo.getOutputsMap());
-          processSizedElementsAndRestrictions.setUniqueName(
-              generateUniquePCollectonName(
-                  splittableParDo.getUniqueName() + "/ProcessSizedElementsAndRestrictions",
-                  existingComponents));
-          processSizedElementsAndRestrictions.setSpec(
-              FunctionSpec.newBuilder()
-                  .setUrn(
-                      PTransformTranslation.SPLITTABLE_PROCESS_SIZED_ELEMENTS_AND_RESTRICTIONS_URN)
-                  .setPayload(splittableParDo.getSpec().getPayload()));
-          processSizedElementsAndRestrictions.setEnvironmentId(splittableParDo.getEnvironmentId());
+        if (!isDrain) {
```

Comment:
Should we able to simplify this more with?
```
PTransform.Builder newCompositeRoot = ... add pair w/ restriction and split and size ...
String processSizedElementsInputPCollectionId = splitAndSizeOutId;
if (isDrain) {
  ... add drain transform ...
  newCompositeRoot.add drain transform
  String processSizedElementsInputPCollectionId = truncateAndSizeOutId;
}
... add process sized elements transform ...
newCompositeRoot.add process sized transform
```

I don't think we should need two copies of adding `ProcessSizedElementsAndRestrictions`